scvi/external/sysvi/_module.py

```diff
@@ -0,0 +1,445 @@
+        # Generative
+        selected_system = self.random_select_systems(tensors["system"])
+        generative_inputs = self._get_generative_input(
+            tensors,
+            inference_outputs,
+            selected_system=selected_system,
+            **get_generative_input_kwargs,
+        )
+        generative_outputs = self.generative(
+            **generative_inputs, x_x=True, x_y=True, **generative_kwargs
+        )
+        # Inference cycle
+        inference_cycle_inputs = self._get_inference_cycle_input(
+            tensors=tensors,
+            generative_outputs=generative_outputs,
+            selected_system=selected_system,
+            **get_inference_input_kwargs,
+        )
+        inference_cycle_outputs = self.inference(**inference_cycle_inputs, **inference_kwargs)
+
+        # Combine outputs of all forward pass components
+        inference_outputs_merged = dict(**inference_outputs)
+        inference_outputs_merged.update(
+            **{k.replace("z", "z_cyc"): v for k, v in inference_cycle_outputs.items()}
+        )
+        generative_outputs_merged = dict(**generative_outputs)
+
+        if compute_loss:
+            losses = self.loss(
+                tensors=tensors,
+                inference_outputs=inference_outputs_merged,
+                generative_outputs=generative_outputs_merged,
+                **loss_kwargs,
+            )
+            return inference_outputs_merged, generative_outputs_merged, losses
+        else:
+            return inference_outputs_merged, generative_outputs_merged
+
+    def loss(
+        self,
+        tensors,
+        inference_outputs,
+        generative_outputs,
+        kl_weight: float = 1.0,
+        reconstruction_weight: float = 1.0,
+        z_distance_cycle_weight: float = 2.0,
+    ):
+        x_true = tensors[REGISTRY_KEYS.X_KEY]
+
+        # Reconstruction loss
+
+        def reconst_loss_part(x_m, x, x_v):
+            """Compute reconstruction loss"""
+            return torch.nn.GaussianNLLLoss(reduction="none")(x_m, x, x_v).sum(dim=1)
+
+        # Reconstruction loss
+        reconst_loss_x = reconst_loss_part(
+            x_m=generative_outputs["x_m"], x=x_true, x_v=generative_outputs["x_v"]
+        )
+        reconst_loss = reconst_loss_x
+
+        # Kl divergence on latent space
+        kl_divergence_z = self.prior.kl(
+            m_q=inference_outputs["z_m"],
+            v_q=inference_outputs["z_v"],
+            z=inference_outputs["z"],
+        )
+
+        def z_dist(z_x_m: torch.Tensor, z_y_m: torch.Tensor):
+            """MSE loss between standardised inputs with standardizer fitted on concatenation of both inputs"""
+            # Standardise data (jointly both z-s) before MSE calculation
+            z = torch.concat([z_x_m, z_y_m])
+            means = z.mean(dim=0, keepdim=True)
+            stds = z.std(dim=0, keepdim=True)
+
+            def standardize(x):
+                return (x - means) / stds
```

Comment:
Does this diverge? Adding clamping here or an offset to stds might be more stable.